File: ElectroRemote/Data/CarRepository.swift
import Foundation

/// Чем закончилась попытка разбудить машину через сервер.
enum WakeResult {
    /// Команда принята; `how` — каким путём машину будили.
    case sent(String)
    /// Сервер ответил отказом или не ответил вовсе.
    case failed(String)
    /// Сервер как путь недоступен: отключён в настройках или нет входа.
    case noServer
}

/// Ошибка с человекочитаемой причиной — для Result из репозитория.
struct RepoError: LocalizedError {
    let message: String
    let code: Int?
    init(_ message: String, code: Int? = nil) { self.message = message; self.code = code }
    var errorDescription: String? { message }
}

/// Единая точка доступа к машине — через наш сервер, и только через него.
final class CarRepository {
    private let settings: Settings
    private let cloud: CloudClient

    init(settings: Settings) {
        self.settings = settings
        self.cloud = CloudClient(settings: settings)
    }

    /// Снять состояние с сервера: он спрашивает машину сам.
    func refresh() async -> CarState {
        if Demo.enabled { return Demo.state() }
        let now = nowMillis()
        var empty = CarState(); empty.updatedAt = now
        guard settings.cloudEnabled, let id = await resolveVehicleId() else { return empty }
        do { return CarState.fromCloud(try await cloud.status(id), now: now) } catch { return empty }
    }

    /// Отправить команду. Путь один — сервер.
    func send(_ cmd: VehicleCommand) async -> CmdResult {
        if Demo.enabled { return .ok }
        guard settings.cloudEnabled else { return .failed(L("Сервер отключён в настройках")) }
        guard settings.loggedIn else { return .failed(L("Вход в аккаунт не выполнен")) }
        guard let id = await resolveVehicleId() else { return .failed(L("Машина не выбрана")) }

        // Замок и климат — своими эндпоинтами; остальное — одной командой кузова.
        do {
            let dto: CommandDto
            switch (cmd.type, cmd.value) {
            case (Cmd.LOCK, "1"): dto = try await cloud.unlock(id)
            case (Cmd.LOCK, "0"): dto = try await cloud.lock(id)
            case (Cmd.AC, "1"): dto = try await cloud.climateOn(id)
            case (Cmd.AC, "0"): dto = try await cloud.climateOff(id)
            default: dto = try await cloud.headControl(id, HeadControlRequest(type: cmd.type, value: cmd.value))
            }
            return Self.outcome(dto)
        } catch {
            return .failed(reasonOf(error))
        }
    }

    /// Включить климат с таймером авто-выключения; nil — умолчание сервера.
    func climateOn(runMinutes: Int?) async -> CmdResult {
        if Demo.enabled { return .ok }
        guard settings.cloudEnabled else { return .failed(L("Сервер отключён в настройках")) }
        guard settings.loggedIn else { return .failed(L("Вход в аккаунт не выполнен")) }
        guard let id = await resolveVehicleId() else { return .failed(L("Машина не выбрана")) }
        var req = CommandRequest()
        if let m = runMinutes { req.params = ["run_minutes": .number(Double(m))] }
        do { return Self.outcome(try await cloud.climateOn(id, req)) } catch { return .failed(reasonOf(error)) }
    }

    /// SUCCESS_UNVERIFIED — тоже успех: голова подтвердила приём.
    private static func outcome(_ dto: CommandDto) -> CmdResult {
        dto.status.hasPrefix("SUCCESS") ? .ok : .failed((dto.error?.isEmpty == false ? dto.error : nil) ?? dto.status)
    }

    // --- пробуждение ---

    /// Контакты поддержки (открытый эндпоинт). nil — не достали.
    func support() async -> SupportDto? { Demo.enabled ? Demo.support : try? await cloud.support() }
    func products() async -> [ProductDto] { (try? await cloud.products()) ?? [] }
    /// Заявка на товар; nil — отправлено, иначе текст ошибки.
    func order(_ req: OrderRequest) async -> String? {
        do { _ = try await cloud.order(req); return nil } catch { return reasonOf(error) }
    }

    /// Файл к отзыву, уже прочитанный в память.
    struct FeedbackFile { let name: String; let mime: String; let data: Data }
    /// Что вышло: сколько вложений долетело и сколько отвалилось.
    struct FeedbackOutcome { let id: String; let attached: Int; let failed: Int }

    /// Отзыв в админку; вложения уходят следом по одному, отзыв без них всё равно отправлен.
    func sendFeedback(kind: String, text: String, appVersion: String, files: [FeedbackFile]) async throws -> FeedbackOutcome {
        if Demo.enabled { return FeedbackOutcome(id: "demo", attached: files.count, failed: 0) }
        let created = try await cloud.sendFeedback(FeedbackRequest(
            kind: kind, text: text, vehicle_id: settings.vehicleId.flatMap { $0.isEmpty ? nil : $0 }, app_version: appVersion))
        var failed = 0
        for f in files {
            do { _ = try await cloud.attachToFeedback(created.id, name: f.name, mime: f.mime, data: f.data) } catch { failed += 1 }
        }
        return FeedbackOutcome(id: created.id, attached: files.count - failed, failed: failed)
    }

    func wake() async -> WakeResult {
        if Demo.enabled { return .sent(L("через пробуждалку в машине")) }
        guard settings.cloudEnabled, settings.loggedIn else { return .noServer }
        guard let id = await resolveVehicleId() else { return .noServer }
        do {
            let dto = try await cloud.wake(id)
            if dto.status.hasPrefix("SUCCESS") { return .sent(channelName(dto.channel)) }
            return .failed((dto.error?.isEmpty == false ? dto.error : nil) ?? dto.status)
        } catch { return .failed(reasonOf(error)) }
    }

    /// Отпустить машину обратно в сон — через ту же железку.
    func sleep() async -> WakeResult {
        if Demo.enabled { return .sent(L("через пробуждалку в машине")) }
        guard settings.cloudEnabled, settings.loggedIn else { return .noServer }
        guard let id = await resolveVehicleId() else { return .noServer }
        do {
            let dto = try await cloud.sleep(id)
            if dto.status.hasPrefix("SUCCESS") { return .sent(channelName(dto.channel)) }
            return .failed((dto.error?.isEmpty == false ? dto.error : nil) ?? dto.status)
        } catch { return .failed(reasonOf(error)) }
    }

    private func channelName(_ channel: String?) -> String {
        switch channel {
        case "WAKE_DEVICE": return L("через пробуждалку в машине")
        case "LEAPMOTOR_CLOUD": return L("через облако")
        default: return L("через сервер")
        }
    }

    /// Причина отказа человеческим языком: из тела 5xx/4xx (`error`/`detail`),
    /// иначе обобщённая сетевая причина без адресов.
    func reasonOf(_ error: Error) -> String {
        if case ApiError.http(_, let body) = error, !body.isEmpty,
           let obj = try? JSONSerialization.jsonObject(with: Data(body.utf8)) as? [String: Any] {
            let text = (obj["error"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                ?? (obj["detail"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            if let text { return scrubAddresses(text) ?? text }
        }
        if let f = friendlyNetworkError(error) { return f }
        if let e = error as? ApiError { return scrubAddresses(e.errorDescription) ?? L("Сервер недоступен") }
        return scrubAddresses(error.localizedDescription) ?? L("Сервер недоступен")
    }

    func reason(_ error: Error) -> String { reasonOf(error) }

    // --- вход в аккаунт ---

    func login(email: String, password: String) async throws {
        if Demo.enabled { settings.token = "demo"; settings.email = email; settings.vehicleId = Demo.vehicle.vehicle_id; return }
        let token = try await cloud.login(LoginRequest(email: email, password: password)).access_token
        settings.token = token
        settings.email = email
        _ = await resolveVehicleId()
    }

    /// Регистрация. Сервер сразу отдаёт токен — второй раз входить не нужно.
    func register(email: String, password: String, name: String?, phone: String?) async throws {
        let created = try await cloud.register(RegisterRequest(
            email: email, password: password,
            name: (name?.isEmpty ?? true) ? nil : name,
            phone: (phone?.isEmpty ?? true) ? nil : phone
        ))
        settings.token = created.access_token
        settings.email = created.user.email
    }

    func requestPasswordReset(email: String, contact: String?) async throws {
        _ = try await cloud.requestPasswordReset(PasswordResetRequest(email: email, contact: (contact?.isEmpty ?? true) ? nil : contact))
    }

    /// Погасить код с QR и получить доступ к машине; она сразу становится выбранной.
    func claimPairing(code: String) async throws -> VehicleDto {
        let v = try await cloud.pairClaim(PairClaimRequest(code: code))
        settings.vehicleId = v.vehicle_id
        return v
    }

    /// Удалить свой аккаунт на сервере (необратимо). После — локальный выход.
    func deleteAccount(password: String) async throws {
        try await cloud.deleteAccount(AccountDeleteRequest(current_password: password))
        settings.logout()
    }

    func news() async -> [NewsItem] { Demo.enabled ? Demo.news : ((try? await cloud.news()) ?? []) }
    func newsPublic() async -> [NewsItem] { Demo.enabled ? Demo.news : ((try? await cloud.newsPublic()) ?? []) }

    func logout() { settings.logout() }

    // --- парк ---

    func vehicles() async throws -> [VehicleDto] { Demo.enabled ? [Demo.vehicle] : try await cloud.vehicles() }

    func isUnauthorized(_ error: Error) -> Bool { (error as? ApiError)?.code == 401 }

    func selectVehicle(_ vehicleId: String) { settings.vehicleId = vehicleId }

    /// vehicle_id из настроек или подбирается сам: из парка выбирается C16.
    private func resolveVehicleId() async -> String? {
        if let id = settings.vehicleId, !id.isEmpty { return id }
        guard settings.loggedIn else { return nil }
        let all = (try? await cloud.vehicles()) ?? []
        let mine = all.first { ($0.model ?? "").caseInsensitiveCompare("C16") == .orderedSame } ?? all.first
        guard let mine else { return nil }
        settings.vehicleId = mine.vehicle_id
        return mine.vehicle_id
    }

    // --- возможности, сцены, расписание, голос ---

    func capabilities() async -> CapabilitiesDto? {
        if Demo.enabled { return Demo.capabilities }
        guard let id = await resolveVehicleId() else { return nil }
        return try? await cloud.capabilities(id)
    }

    func scenes() async -> [SceneTemplateDto] {
        if Demo.enabled { return Demo.scenes }
        guard let id = await resolveVehicleId() else { return [] }
        return (try? await cloud.scenes(id)) ?? []
    }

    func createScene(name: String, steps: [SceneStepDto]) async throws -> SceneTemplateDto {
        guard let id = await resolveVehicleId() else { throw RepoError(L("Машина не выбрана")) }
        return try await cloud.createScene(id, SceneTemplateRequest(name: name, steps: steps))
    }

    func deleteScene(_ templateId: String) async throws {
        guard let id = await resolveVehicleId() else { throw RepoError(L("Машина не выбрана")) }
        try await cloud.deleteScene(id, templateId)
    }

    func runScene(_ templateId: String) async -> CmdResult {
        guard let id = await resolveVehicleId() else { return .failed(L("Машина не выбрана")) }
        do { return Self.outcome(try await cloud.runScene(id, templateId)) } catch { return .failed(reasonOf(error)) }
    }

    func climateSchedules() async -> [ClimateScheduleDto] {
        if Demo.enabled { return Demo.schedules }
        guard let id = await resolveVehicleId() else { return [] }
        return (try? await cloud.climateSchedules(id)) ?? []
    }

    func createClimateSchedule(_ body: ClimateScheduleRequest) async throws -> ClimateScheduleDto {
        guard let id = await resolveVehicleId() else { throw RepoError(L("Машина не выбрана")) }
        return try await cloud.createClimateSchedule(id, body)
    }

    func setScheduleEnabled(_ scheduleId: String, _ enabled: Bool) async throws {
        guard let id = await resolveVehicleId() else { throw RepoError(L("Машина не выбрана")) }
        _ = try await cloud.setScheduleEnabled(id, scheduleId, enabled)
    }

    func deleteClimateSchedule(_ scheduleId: String) async throws {
        guard let id = await resolveVehicleId() else { throw RepoError(L("Машина не выбрана")) }
        try await cloud.deleteClimateSchedule(id, scheduleId)
    }

    /// Отвязать машину от аккаунта. Сама машина остаётся, снимается только доступ.
    func unlinkVehicle(_ vehicleId: String) async throws { try await cloud.unlinkVehicle(vehicleId) }

    func voiceIntents() async -> [VoiceIntentDto] { Demo.enabled ? Demo.voice : ((try? await cloud.voiceIntents()) ?? []) }

    func runVoice(_ intent: String) async -> CmdResult {
        guard let id = await resolveVehicleId() else { return .failed(L("Машина не выбрана")) }
        do { return Self.outcome(try await cloud.voice(id, intent)) } catch { return .failed(reasonOf(error)) }
    }
}

/// Разбор QR с экрана машины: `electro://pair?c=<код>&h=<адрес головы>`.
/// Параметр `h` остался от прежней схемы и игнорируется.
enum Pairing {
    static func parse(_ payload: String) -> String? {
        guard let comps = URLComponents(string: payload.trimmingCharacters(in: .whitespacesAndNewlines)),
              comps.scheme == "electro" else { return nil }
        return comps.queryItems?.first(where: { $0.name == "c" })?.value
    }

    /// Привязывает машину по отсканированному коду; результат — текст для показа.
    static func claim(_ payload: String, repo: CarRepository) async throws -> String {
        guard let code = parse(payload) else { throw RepoError(L("Это не QR машины Electro")) }
        if code.trimmingCharacters(in: .whitespaces).isEmpty { throw RepoError(L("В коде нет кода привязки")) }
        do {
            let v = try await repo.claimPairing(code: code)
            return L("Машина привязана: {0}", v.name)
        } catch {
            throw RepoError(repo.reason(error), code: (error as? ApiError)?.code)
        }
    }
}

File: android/app/src/main/java/uz/electro/remote/data/CarRepository.kt
package uz.electro.remote.data

import uz.electro.remote.i18n.S
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import retrofit2.HttpException

/** Чем закончилась попытка разбудить машину через сервер. */
sealed interface WakeResult {
    /** Команда принята; [how] — каким путём машину будили. */
    data class Sent(val how: String) : WakeResult

    /** Сервер ответил отказом или не ответил вовсе. */
    data class Failed(val reason: String) : WakeResult

    /** Сервер как путь недоступен: отключён в настройках или нет входа. */
    data object NoServer : WakeResult
}

/**
 * Единая точка доступа к машине — через наш сервер, и только через него.
 *
 * Прямого канала на голову больше нет. Он требовал, чтобы телефон и машина
 * сидели в одной сети, а на голове был включён adb — в боевой машине его
 * выключают сразу после установки. Один путь вместо двух заодно снимает
 * разнобой: раньше набор команд и скорость опроса зависели от того, каким
 * каналом повезло пройти.
 */
/** Файл к отзыву, уже прочитанный в память: имя для админки, тип, байты. */
class FeedbackFile(val name: String, val mime: String, val bytes: ByteArray)

/** Что вышло: id отзыва и сколько вложений долетело/отвалилось. */
data class FeedbackOutcome(val id: String, val attached: Int, val failed: Int)

class CarRepository(private val settings: Settings) {

    private val cloud = CloudClient(settings)

    /** Снять состояние с сервера: он спрашивает машину сам. */
    suspend fun refresh(): CarState = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis()
        if (!settings.cloudEnabled) return@withContext CarState(updatedAt = now)
        val id = resolveVehicleId() ?: return@withContext CarState(updatedAt = now)
        runCatching { cloud.api.status(id) }
            .map { CarState.fromCloud(it, now) }
            .getOrElse { CarState(updatedAt = now) }
    }

    /** Отправить команду. Путь один — сервер. */
    suspend fun send(cmd: VehicleCommand): CmdResult = withContext(Dispatchers.IO) {
        sendViaCloud(cmd)
    }

    /**
     * Включить климат с таймером авто-выключения (GWM «время работы»).
     *
     * `runMinutes` уходит серверу в params: 0 — без таймера, null — умолчание
     * сервера. Гасит забытый климат сам сервер, а не телефон.
     */
    suspend fun climateOn(runMinutes: Int?): CmdResult = withContext(Dispatchers.IO) {
        if (!settings.cloudEnabled) return@withContext CmdResult.Failed(S("Сервер отключён в настройках"))
        if (!settings.loggedIn) return@withContext CmdResult.Failed(S("Вход в аккаунт не выполнен"))
        val id = resolveVehicleId() ?: return@withContext CmdResult.Failed(S("Машина не выбрана"))
        val params = if (runMinutes != null) mapOf("run_minutes" to runMinutes) else emptyMap()
        runCatching { cloud.api.climateOn(id, CommandRequest(params = params)) }
            .map {
                if (it.status.startsWith("SUCCESS")) CmdResult.Ok
                else CmdResult.Failed(it.error ?: it.status)
            }
            .getOrElse { CmdResult.Failed(reasonOf(it)) }
    }

    private suspend fun sendViaCloud(cmd: VehicleCommand): CmdResult {
        if (!settings.cloudEnabled) return CmdResult.Failed(S("Сервер отключён в настройках"))
        if (!settings.loggedIn) return CmdResult.Failed(S("Вход в аккаунт не выполнен"))
        val id = resolveVehicleId() ?: return CmdResult.Failed(S("Машина не выбрана"))

        // Замок и климат — своими эндпоинтами: у них на сервере отдельные права
        // и своя проверка. Всё остальное уходит одной командой кузова: карту
        // типов знает голова, серверу её знать незачем.
        val call: (suspend () -> CommandDto) = when {
            cmd.type == Cmd.LOCK && cmd.value == "1" -> ({ cloud.api.unlock(id, CommandRequest()) })
            cmd.type == Cmd.LOCK && cmd.value == "0" -> ({ cloud.api.lock(id, CommandRequest()) })
            cmd.type == Cmd.AC && cmd.value == "1" -> ({ cloud.api.climateOn(id, CommandRequest()) })
            cmd.type == Cmd.AC && cmd.value == "0" -> ({ cloud.api.climateOff(id, CommandRequest()) })
            else -> ({ cloud.api.headControl(id, HeadControlRequest(cmd.type, cmd.value)) })
        }

        // SUCCESS_UNVERIFIED — тоже успех: голова подтвердила приём, а состояние
        // читается отдельно и к команде не привязано.
        return runCatching { call.invoke() }
            .map {
                if (it.status.startsWith("SUCCESS")) CmdResult.Ok
                else CmdResult.Failed(it.error ?: it.status)
            }
            .getOrElse { CmdResult.Failed(reasonOf(it)) }
    }

    // --- пробуждение ------------------------------------------------------

    /**
     * Разбудить машину через сервер.
     *
     * Сервер сам выбирает путь — пробуждалку в машине или облако — и, в отличие
     * от SMS, отвечает, дошла ли команда: у SMS обратной связи нет вовсе, и
     * «отправлено» там означало лишь «отдано оператору».
     */
    /** Контакты поддержки (открытый эндпоинт, ключ не нужен). null — не достали. */
    suspend fun support(): SupportDto? = withContext(Dispatchers.IO) {
        runCatching { cloud.api.support() }.getOrNull()
    }

    /**
     * Отзыв в админку: замечание, идея, вопрос. Машина — текущая, если выбрана.
     * Вложения (скриншоты/записи экрана) уходят следом по одному; если какое-то
     * не долетело — отзыв всё равно отправлен, об этом говорит результат.
     */
    suspend fun sendFeedback(
        kind: String, text: String, appVersion: String,
        attachments: List<FeedbackFile> = emptyList(),
    ): Result<FeedbackOutcome> = withContext(Dispatchers.IO) {
        runCatching {
            val created = cloud.api.sendFeedback(FeedbackRequest(
                kind = kind, text = text,
                vehicle_id = settings.vehicleId?.takeIf { it.isNotBlank() },
                app = "phone", app_version = appVersion,
            ))
            var failed = 0
            for (f in attachments) {
                val body = f.bytes.toRequestBody(f.mime.toMediaTypeOrNull())
                runCatching { cloud.api.attachToFeedback(created.id, f.name, body) }
                    .onFailure { failed++ }
            }
            FeedbackOutcome(created.id, attachments.size - failed, failed)
        }
    }

    suspend fun wake(): WakeResult = withContext(Dispatchers.IO) {
        if (!settings.cloudEnabled || !settings.loggedIn) return@withContext WakeResult.NoServer
        val id = resolveVehicleId() ?: return@withContext WakeResult.NoServer
        runCatching { cloud.api.wake(id, CommandRequest()) }.fold(
            onSuccess = { dto ->
                // SUCCESS_UNVERIFIED — тоже успех: команда принята, а проверить,
                // что машина проснулась, этим каналом нечем.
                if (dto.status.startsWith("SUCCESS")) WakeResult.Sent(channelName(dto.channel))
                else WakeResult.Failed(dto.error?.ifBlank { null } ?: dto.status)
            },
            onFailure = { WakeResult.Failed(reasonOf(it)) },
        )
    }

    /**
     * Причина отказа человеческим языком.
     *
     * Непрошедшую команду backend отдаёт кодом 5xx, и без разбора тела на экран
     * попадало бы «HTTP 502» вместо «пробуждалка не ответила за 15 с» — то есть
     * ровно то, что нужно знать, оставалось бы в логах сервера.
     */
    private fun reasonOf(error: Throwable): String {
        val body = (error as? HttpException)?.response()?.errorBody()?.string().orEmpty()
        if (body.isNotBlank()) {
            runCatching {
                val json = JSONObject(body)
                val text = json.optString("error").ifBlank { json.optString("detail") }
                if (text.isNotBlank()) return scrubAddresses(text) ?: text
            }
        }
        // Сетевые ошибки OkHttp несут в тексте хост/адрес («Unable to resolve
        // host leapmotor.evon.uz», «Failed to connect to /62.171.159.63») — их
        // на экран показывать нельзя. Переводим в обобщённую причину.
        friendlyNetworkError(error)?.let { return it }
        return scrubAddresses(error.message) ?: S("Сервер недоступен")
    }

    /**
     * Отпустить машину обратно в сон — через ту же железку.
     *
     * Запасного пути здесь нет: у облака нет такого понятия, а SMS уходит
     * вслепую. Не прошло — значит не прошло.
     */
    suspend fun sleep(): WakeResult = withContext(Dispatchers.IO) {
        if (!settings.cloudEnabled || !settings.loggedIn) return@withContext WakeResult.NoServer
        val id = resolveVehicleId() ?: return@withContext WakeResult.NoServer
        runCatching { cloud.api.sleep(id, CommandRequest()) }.fold(
            onSuccess = { dto ->
                if (dto.status.startsWith("SUCCESS")) WakeResult.Sent(channelName(dto.channel))
                else WakeResult.Failed(dto.error?.ifBlank { null } ?: dto.status)
            },
            onFailure = { WakeResult.Failed(reasonOf(it)) },
        )
    }

    private fun channelName(channel: String?): String = when (channel) {
        "WAKE_DEVICE" -> S("через пробуждалку в машине")
        "LEAPMOTOR_CLOUD" -> S("через облако")
        else -> S("через сервер")
    }

    // --- вход в аккаунт --------------------------------------------------

    suspend fun login(email: String, password: String): Result<Unit> = withContext(Dispatchers.IO) {
        runCatching {
            val token = cloud.api.login(LoginRequest(email, password)).access_token
            settings.token = token
            settings.email = email
            resolveVehicleId()
            Unit
        }
    }

    /**
     * Регистрация. Учётка общая с сайтом: тот же backend, та же таблица людей.
     *
     * Сервер сразу отдаёт токен, поэтому повторный вход не нужен — сохраняем
     * сессию так же, как после login().
     */
    suspend fun register(
        email: String, password: String, name: String?, phone: String?,
    ): Result<Unit> = withContext(Dispatchers.IO) {
        runCatching {
            val created = cloud.api.register(RegisterRequest(
                email = email,
                password = password,
                name = name?.takeIf { it.isNotBlank() },
                phone = phone?.takeIf { it.isNotBlank() },
            ))
            settings.token = created.access_token
            settings.email = created.user.email
            Unit
        }
    }

    /**
     * «Забыл пароль»: заявка мастеру. Почтового канала у установки нет, новый
     * пароль выдаёт и сообщает человек.
     */
    suspend fun requestPasswordReset(email: String, contact: String?): Result<Unit> =
        withContext(Dispatchers.IO) {
            runCatching {
                cloud.api.requestPasswordReset(
                    PasswordResetRequest(email, contact?.takeIf { it.isNotBlank() })
                )
                Unit
            }
        }

    /**
     * Погасить код с QR на экране машины и получить доступ к ней.
     *
     * Привязанную машину сразу делаем выбранной: она у аккаунта первая и,
     * скорее всего, единственная, и заставлять выбирать её вручную незачем.
     */
    suspend fun claimPairing(code: String): Result<VehicleDto> = withContext(Dispatchers.IO) {
        runCatching { cloud.api.pairClaim(PairClaimRequest(code)) }
            .onSuccess { settings.vehicleId = it.vehicle_id }
    }

    /** Причина отказа человеческим языком — та же, что и у команд. */
    fun reason(error: Throwable): String = reasonOf(error)

    /** Удалить свой аккаунт на сервере (необратимо). После — локальный выход. */
    suspend fun deleteAccount(password: String): Result<Unit> = withContext(Dispatchers.IO) {
        runCatching {
            val r = cloud.api.deleteAccount(AccountDeleteRequest(password))
            if (!r.isSuccessful) throw retrofit2.HttpException(r)
            settings.logout()
            Unit
        }
    }

    suspend fun news(): List<NewsItemDto> = withContext(Dispatchers.IO) {
        runCatching { cloud.api.news() }.getOrElse { emptyList() }
    }

    /** Push-токен серверу; не критично — при неудаче повторим при следующем запуске. */
    suspend fun registerPushToken(token: String): Boolean = withContext(Dispatchers.IO) {
        if (!settings.loggedIn) return@withContext false
        runCatching { cloud.api.registerPushToken(PushTokenBody(token = token)).isSuccessful }.getOrDefault(false)
    }

    suspend fun removePushToken(token: String) = withContext(Dispatchers.IO) {
        runCatching { cloud.api.removePushToken(PushTokenRemove(token)) }
    }

    fun logout() = settings.logout()

    // --- парк -------------------------------------------------------------

    /** Машины, к которым допущен аккаунт. */
    suspend fun vehicles(): Result<List<VehicleDto>> = withContext(Dispatchers.IO) {
        runCatching { cloud.api.vehicles() }
    }

    /** Отказ по недействительной сессии — его нельзя показывать как ошибку сети. */
    fun isUnauthorized(error: Throwable): Boolean =
        (error as? HttpException)?.code() == 401

    /** Выбранная машина: с этого момента команды и статус идут в неё. */
    fun selectVehicle(vehicleId: String) {
        settings.vehicleId = vehicleId
    }

    /**
     * vehicle_id известен из настроек или подбирается сам.
     *
     * Из парка выбирается C16: это приложение управляет ею, а у C11 и C01 свой
     * клиент и другой набор команд. Взять просто первую машину значило бы
     * отправлять команды C16 в ту, которая их не понимает.
     */
    private suspend fun resolveVehicleId(): String? {
        settings.vehicleId?.takeIf { it.isNotBlank() }?.let { return it }
        if (!settings.loggedIn) return null
        val all = runCatching { cloud.api.vehicles() }.getOrNull().orEmpty()
        val mine = all.firstOrNull { it.model.equals("C16", ignoreCase = true) }
            ?: all.firstOrNull() ?: return null
        settings.vehicleId = mine.vehicle_id
        return mine.vehicle_id
    }

    // --- возможности, сцены, расписание ----------------------------------

    /** Что рисовать: набор кнопок решает сервер, а не приложение. */
    suspend fun capabilities(): CapabilitiesDto? = withContext(Dispatchers.IO) {
        val id = resolveVehicleId() ?: return@withContext null
        runCatching { cloud.api.capabilities(id) }.getOrNull()
    }

    suspend fun scenes(): List<SceneTemplateDto> = withContext(Dispatchers.IO) {
        val id = resolveVehicleId() ?: return@withContext emptyList()
        runCatching { cloud.api.scenes(id) }.getOrElse { emptyList() }
    }

    suspend fun createScene(name: String, steps: List<SceneStepDto>): Result<SceneTemplateDto> =
        withContext(Dispatchers.IO) {
            val id = resolveVehicleId()
                ?: return@withContext Result.failure(IllegalStateException(S("Машина не выбрана")))
            runCatching { cloud.api.createScene(id, SceneTemplateRequest(name, steps)) }
        }

    suspend fun deleteScene(templateId: String): Result<Unit> = withContext(Dispatchers.IO) {
        val id = resolveVehicleId()
            ?: return@withContext Result.failure(IllegalStateException(S("Машина не выбрана")))
        runCatching { cloud.api.deleteScene(id, templateId); Unit }
    }

    suspend fun runScene(templateId: String): CmdResult = withContext(Dispatchers.IO) {
        val id = resolveVehicleId() ?: return@withContext CmdResult.Failed(S("Машина не выбрана"))
        runCatching { cloud.api.runScene(id, templateId, CommandRequest()) }
            .map {
                if (it.status.startsWith("SUCCESS")) CmdResult.Ok
                else CmdResult.Failed(it.error ?: it.status)
            }
            .getOrElse { CmdResult.Failed(reasonOf(it)) }
    }

    suspend fun climateSchedules(): List<ClimateScheduleDto> = withContext(Dispatchers.IO) {
        val id = resolveVehicleId() ?: return@withContext emptyList()
        runCatching { cloud.api.climateSchedules(id) }.getOrElse { emptyList() }
    }

    suspend fun createClimateSchedule(body: ClimateScheduleRequest): Result<ClimateScheduleDto> =
        withContext(Dispatchers.IO) {
            val id = resolveVehicleId()
                ?: return@withContext Result.failure(IllegalStateException(S("Машина не выбрана")))
            runCatching { cloud.api.createClimateSchedule(id, body) }
        }

    suspend fun setScheduleEnabled(scheduleId: String, enabled: Boolean): Result<Unit> =
        withContext(Dispatchers.IO) {
            val id = resolveVehicleId()
                ?: return@withContext Result.failure(IllegalStateException(S("Машина не выбрана")))
            runCatching { cloud.api.setScheduleEnabled(id, scheduleId, EnabledRequest(enabled)); Unit }
        }

    suspend fun deleteClimateSchedule(scheduleId: String): Result<Unit> =
        withContext(Dispatchers.IO) {
            val id = resolveVehicleId()
                ?: return@withContext Result.failure(IllegalStateException(S("Машина не выбрана")))
            runCatching { cloud.api.deleteClimateSchedule(id, scheduleId); Unit }
        }

    /** Отвязать машину от аккаунта. Сама машина остаётся, снимается только доступ. */
    suspend fun unlinkVehicle(vehicleId: String): Result<Unit> = withContext(Dispatchers.IO) {
        runCatching { cloud.api.unlinkVehicle(vehicleId); Unit }
    }

    suspend fun voiceIntents(): List<VoiceIntentDto> = withContext(Dispatchers.IO) {
        runCatching { cloud.api.voiceIntents() }.getOrElse { emptyList() }
    }

    suspend fun runVoice(intent: String): CmdResult = withContext(Dispatchers.IO) {
        val id = resolveVehicleId() ?: return@withContext CmdResult.Failed(S("Машина не выбрана"))
        runCatching { cloud.api.voice(id, VoiceRequestBody(intent)) }
            .map {
                if (it.status.startsWith("SUCCESS")) CmdResult.Ok
                else CmdResult.Failed(it.error ?: it.status)
            }
            .getOrElse { CmdResult.Failed(reasonOf(it)) }
    }
}
